e2e/e2e_test.go

```diff
@@ -0,0 +1,173 @@
+	"github.com/ethereum/go-ethereum/common"
+)
+
+const TestTimeout = time.Second * 30
+const TestEthEndpoint = "ws://localhost:8545"
+const TestSubEndpoint = "ws://localhost:9944"
+const EthChainId = msg.ChainId(0)
+const SubChainId = msg.ChainId(1)
+
+var AliceEthKp = keystore.TestKeyRing.EthereumKeys[keystore.AliceKey]
+var CharlieEthKp = keystore.TestKeyRing.EthereumKeys[keystore.BobKey]
+var AliceSubKp = keystore.TestKeyRing.SubstrateKeys[keystore.AliceKey]
+var BobSubKp = keystore.TestKeyRing.SubstrateKeys[keystore.BobKey]
+var CharlieEthAddr = common.HexToAddress(CharlieEthKp.Address())
+
+func setLogger(lvl log.Lvl) {
+	logger := log.Root()
+	handler := logger.GetHandler()
+	log.Root().SetHandler(log.LvlFilterHandler(lvl, handler))
+}
+
+func createEthConfig(key, bridgeAddress, erc20HandlerAddress, genericHandler string) *core.ChainConfig {
+	return &core.ChainConfig{
+		Name:         fmt.Sprintf("ethereum(%s)", key),
+		Id:           EthChainId,
+		Endpoint:     TestEthEndpoint,
+		From:         "",
+		KeystorePath: key,
+		Insecure:     true,
+		Opts: map[string]string{
+			"contract":       bridgeAddress,
+			"erc20Handler":   erc20HandlerAddress,
+			"genericHandler": genericHandler,
+		},
+	}
+}
+
+func createSubConfig(key string) *core.ChainConfig {
+	return &core.ChainConfig{
+		Name:         fmt.Sprintf("substrate(%s)", key),
+		Id:           SubChainId,
+		Endpoint:     TestSubEndpoint,
+		From:         "",
+		KeystorePath: key,
+		Insecure:     true,
+		Opts:         map[string]string{},
+	}
+}
+
+func createAndStartBridge(t *testing.T, name, bridgeAddress, erc20HandlerAddres, genericHandlerAddress string) *core.Core {
+	eth, err := ethereum.InitializeChain(createEthConfig(name, bridgeAddress, erc20HandlerAddres, genericHandlerAddress))
+	if err != nil {
+		t.Fatal(err)
+	}
+	sub, err := substrate.InitializeChain(createSubConfig(name))
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	bridge := core.NewCore()
+	bridge.AddChain(eth)
+	bridge.AddChain(sub)
+
+	err = eth.Start()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	err = sub.Start()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	return bridge
+}
+
+func TestErc20ToSubstrate(t *testing.T) {
+	setLogger(log.LvlInfo)
+
+	// Deploy contracts, mint, approve
+	contracts := deployTestContracts(t, EthChainId)
+	ethClient, opts := createEthClient(t)
+	erc20Contract := deployMintApproveErc20(t, ethClient, opts, contracts.ERC20HandlerAddress)
+
+	// Create and start two bridges with both chains
+	_ = createAndStartBridge(t, "alice", contracts.BridgeAddress.Hex(), contracts.ERC20HandlerAddress.Hex(), contracts.CentrifugeHandlerAddress.Hex())
+	_ = createAndStartBridge(t, "bob", contracts.BridgeAddress.Hex(), contracts.ERC20HandlerAddress.Hex(), contracts.CentrifugeHandlerAddress.Hex())
+
+	// Initiate transfer
+	createErc20Deposit(t, ethClient, opts, contracts, erc20Contract)
+
+	// Check for success event
+	subClient := createSubClient(t, AliceSubKp.AsKeyringPair())
```

Comment:
Is there a way to query substrate to see the balance change?

Reply:
We should probably do asserts on the balances of the accounts

Reply:
This is hard because you have to take fees into account